nomad/structs/config/tls.go

```diff
@@ -38,9 +49,79 @@ type TLSConfig struct {
 	VerifyHTTPSClient bool `mapstructure:"verify_https_client"`
 }
 
+type KeyLoader struct {
+	cacheLock   sync.Mutex
+	certificate *tls.Certificate
+}
+
+// LoadKeyPair reloads the TLS certificate based on the specified certificate
+// and key file. If successful, stores the certificate for further use.
+func (k *KeyLoader) LoadKeyPair(certFile, keyFile string) (*tls.Certificate, error) {
+	k.cacheLock.Lock()
+	defer k.cacheLock.Unlock()
+
+	// Allow downgrading
+	if certFile == "" && keyFile == "" {
+		k.certificate = nil
+		return nil, nil
+	}
+
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load cert/key pair: %v", err)
+	}
+
+	k.certificate = &cert
+	return k.certificate, nil
+}
+
+// GetOutgoingCertificate fetches the currently-loaded certificate. This
+// currently does not consider information in the ClientHello and only returns
+// the certificate that was last loaded.
+func (k *KeyLoader) GetOutgoingCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	k.cacheLock.Lock()
+	defer k.cacheLock.Unlock()
+
+	return k.certificate, nil
+}
+
+// GetKeyLoader returns the keyloader for a TLSConfig object. If the keyloader
+// has not been initialized, it will first do so.
+func (t *TLSConfig) GetKeyLoader() *KeyLoader {
+	t.keyloaderLock.Lock()
+	defer t.keyloaderLock.Unlock()
+
+	// If the keyloader has not yet been initialized, do it here
+	if t.KeyLoader == nil {
+		t.KeyLoader = &KeyLoader{}
+	}
+	return t.KeyLoader
+}
+
+// Copy copies the fields of TLSConfig to another TLSConfig object. Required as
+// to not copy mutexes between objects.
+func (t *TLSConfig) Copy() *TLSConfig {
+
+	new := &TLSConfig{}
+	new.EnableHTTP = t.EnableHTTP
```

Comment:
`*new = *t` is the pattern for getting all non-pointer values.

Reply:
This will also copy lock values, which is a warning in govet. If there is another way that is more idiomatic that doesn't copy lock values, would definitely prefer that.